module/VuFind/src/VuFind/Db/Table/ExternalSession.php

```diff
@@ -0,0 +1,149 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Db_Table
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+namespace VuFind\Db\Table;
+use Zend\Db\Sql\Expression;
+
+/**
+ * Table Definition for external_session
+ *
+ * @category VuFind
+ * @package  Db_Table
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class ExternalSession extends Gateway
+{
+    /**
+     * Constructor
+     */
+    public function __construct()
+    {
+        parent::__construct('external_session', 'VuFind\Db\Row\ExternalSession');
+    }
+
+    /**
+     * Add a mapping between local and external session id's
+     *
+     * @param string $localSessionId    Local (VuFind) session id
+     * @param string $externalSessionId External session id
+     *
+     * @return void
+     */
+    public function addSessionMapping($localSessionId, $externalSessionId)
+    {
+        $this->destroySession($localSessionId);
+        $row = $this->createRow();
+        $row->session_id = $localSessionId;
+        $row->external_session_id = $externalSessionId;
+        $row->created = date('Y-m-d H:i:s');
+        $row->save();
+    }
+
+    /**
+     * Retrieve an object from the database based on an external session ID
+     *
+     * @param string $sid External session ID to retrieve
+     *
+     * @return \VuFind\Db\Row\ExternalSession
+     */
+    public function getByExternalSessionId($sid)
+    {
+        return $this->select(['external_session_id' => $sid])->current();
+    }
+
+    /**
+     * Destroy data for the given session ID.
+     *
+     * @param string $sid Session ID to erase
+     *
+     * @return void
+     */
+    public function destroySession($sid)
+    {
+        $this->delete(['session_id' => $sid]);
+    }
+
+    /**
+     * Delete expired sessions. Allows setting of 'from' and 'to' ID's so that rows
+     * can be deleted in small batches.
+     *
+     * @param int $daysOld Age in days of an "expired" session.
+     * @param int $idFrom  Lowest id of rows to delete.
+     * @param int $idTo    Highest id of rows to delete.
+     *
+     * @return int Number of rows deleted
+     */
+    public function deleteExpired($daysOld = 2, $idFrom = null, $idTo = null)
```

Comment:
This appears to be largely copied-and-pasted from the Search and Session functionality. Might it make sense to implement an ExpirationTrait to reduce the redundancy? (I realize that some callback details need to vary between database tables, but this might be accomplished by including an abstract callback method in the trait that must be implemented by classes using the trait).